torch/amp/autocast_mode.py

```diff
@@ -355,10 +355,25 @@ def __enter__(self):
         torch.autocast_increment_nesting()
         torch.set_autocast_cache_enabled(self._cache_enabled)
 
+        if torch._C._is_torch_function_mode_enabled():
+            return torch.overrides.handle_torch_function(
+                torch.amp._enter_autocast,
```

Comment:
instead of overring the `_enter/exit_autocast` calls in python here, can we just override the single `_set_autocast_enabled(bool)` function from C++, similar to what we do for `_set_autograd_enabled` here? https://github.com/pytorch/pytorch/blob/9b2e453e246ea44f2071dfdac86ec4b0037a51a5/torch/csrc/autograd/init.cpp#L877

Reply:
cc @zou3519 / @albanD in case you have any thoughts/preference. The context is that we need a __torch_function__ override for autocast in order to preserve autocast for non-strict, pre-dispatch export.

And we have two options:

(1) override all 3 mutable autocast API's:
```
torch.set_autocast_enabled
torch.set_autocast_dtype
torch.set_autocast_cache_enabled
```

(2) override the enter/exit API's:
```
torch.amp._enter_autocast
torch.amp._exit_autocast
```

(1) will be a little more in-line with how we override `torch.set_grad_enabled` today (when desugaring `no_grad/enable_grad`). Although it's a bit silly because there are 3 individual mutable autocast API's we'll need to expose to `__torch_function__` instead of 2. If we do (1),  @yushangdi also pointed out that dynamo currently does (2) today, so we'll probably want to fix up dynamo to desugar to (1) for consistency between strict/nonstrict

(2) is what this PR does today.

(I think either is fine, although if we choose one now and want to switch later, there is a bit of BC risk for `__torch_function__` authors)

Reply:
Summarizing chat content:

We could potentially override the C++ API of the following, 

torch.set_autocast_enabled
torch.set_autocast_dtype
torch.set_autocast_cache_enabled

but then it would make strict and non-strict export behavior different. Specifically, in the strict mode produce a single node torch.amp.autocast_mode._enter_autocast (or exit) here: https://github.com/pytorch/pytorch/blob/487873f7cafeb0fd390eaefe40496b804bceabbd/torch/_dynamo/variables/ctx_manager.py#L808

If we override the C++ API, then we would produce three nodes instead. It would also make subsequent pattern matching more complicated.

I guess autocast is a little different from grad_enbaled in that set_autocast_enabled itself is not equivalent to _enter_autocast. We also need a few other operations.

Reply:
minor comment: for option 1, I think we might need to override two additional APIs: 

torch.autocast_increment_nesting()
torch.clear_autocast_cache()

Reply:
torch_function is about overriding the public PyTorch API, so my take is we shouldn't do (2)

Reply:
> set_autocast_enabled

@zou3519 @bdhirsh  I found that `set_autocast_enabled` is a public function that cannot be overridden by ``__torch_function__``. The docstring says this is because none of the arguments of these functions are tensors or tensor-likes.

Does this matter to what we are dicussing?

https://github.com/pytorch/pytorch/blob/main/torch/overrides.py#L100C12-L100C81

https://github.com/pytorch/pytorch/blob/main/torch/overrides.py#L269

Reply:
Do you think we can do something like below (might be hacky):

We only go to `handle_torch_function` if we have PreDispatchTorchFunctionMode on the functional mode stack and we only dispatch to PreDispatchTorchFunctionMode's torch_function, not any other FunctionMode.

```python
	
        if torch._C._is_torch_function_mode_enabled():
            from torch.fx.experimental.proxy_tensor import PreDispatchTorchFunctionMode

            mode_stack = torch.overrides._get_current_function_mode_stack()
            if any(
                isinstance(mode, PreDispatchTorchFunctionMode) for mode in mode_stack
            ):
                return torch.overrides.handle_torch_function(
                    torch.amp._exit_autocast, [], self
                )
```

Reply:
> I found that set_autocast_enabled is a public function that cannot be overridden by __torch_function__ (similarly with the other autocast functions). The docstring says this is because none of the arguments of these functions are tensors or tensor-likes.

This is true for torch-function based tensor-subclasses, but not for modes (which can intercept factory-function-style torch ops). If you look at how we handle `set_autograd_enabled` ([link](https://github.com/pytorch/pytorch/blob/9b2e453e246ea44f2071dfdac86ec4b0037a51a5/torch/csrc/autograd/init.cpp#L875)) - we check that a `TorchFunctionMode` is enabled before dispatching to `handle_torch_function()`

Reply:
Option 1 with a mode sounds like the most straighforward thing to do. An alternative is that we do option 2, but we expose those APIs to users (or figure out what we do want to expose to users).

Reply:
> Option 1 with a mode sounds like the most straighforward thing to do. An alternative is that we do option 2, but we expose those APIs to users (or figure out what we do want to expose to users).

I changed to Option 1 with a mode now. Thanks!